### PR TITLE
Explicit logging when fork network graph is stale

### PR DESCRIPTION
**Pull request by justinharringa**

### CODE REVIEW

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/repository/GitHub.java [hunk 1]
```diff
@@ -24,4 +27,32 @@ public static boolean shouldNotProcessDockerfilesInRepo(Multimap<String, String>
...
+     * @param fork the fork associated with parent
+     * @return the fork is stale for some reason
+     * @throws IOException general issues trying to get information about the parent's branch will throw an IOException
+     */
+    public static boolean isForkStale(GHRepository parent, GHRepository fork) throws IOException {
+        String sha1 = parent.getBranch(parent.getDefaultBranch()).getSHA1();
+        try {
+            fork.getTree(sha1);
+        } catch (GHFileNotFoundException e) {
+            log.info("fork {} is stale... A GitHub administrator will need to rebuild the network graph.",
```
Comment (afalko) [inline dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/repository/GitHub.java:52]: Is there a procedure for doing that?

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java [hunk 2]
```diff
@@ -48,7 +49,18 @@ public GHRepository getOrCreateFork(GHRepository parent) {
...
                 if (thisUserIsOwner(fork)) {
-                    log.info("Fork exists, retrieving full info: {}", fork);
+                    // Check to see if latest sha from parent is in the tree of the fork. If not, it is stale.
+                    // If stale, delete the fork. If we don't do that, we get a 422 Reference update failed
+                    // https://docs.github.com/enterprise/2.22/rest/reference/git#create-a-reference
+                    // Updating the default branch of the fork to the sha results in 422 Object does not exist
+                    // https://docs.github.com/enterprise/2.22/rest/reference/git#update-a-reference
+                    // Confirmed that this does not occur for another similarly old fork
+                    if (GitHub.isForkStale(parent, fork)) {
+                        log.info("Fork's commit graph is inconsistent, you'll likely see a 422 error. Fork info: {}",
```
[the remark this answers is not in the document]
Comment (afalko) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java:59]: Should this be a log.warn or log.error?

**Review by afalko — CHANGES_REQUESTED** [review]
Good change, just a few minor questions.

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/repository/GitHub.java (hunk 1 above)
Comment (justinharringa) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/repository/GitHub.java:54]: Yeah, via admin user interface only unfortunately.

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (justinharringa) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java]: yeah, good point. Switching both

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java [hunk 3]
```diff
@@ -48,7 +49,20 @@ public GHRepository getOrCreateFork(GHRepository parent) {
...
+                    // Check to see if latest sha from parent is in the tree of the fork. If not,
+                    // it is stale. If stale, an administrator will need to rebuild the commit
+                    // graph. If they don't do that, we get a 422 Reference update failed
+                    // https://docs.github.com/enterprise/2.22/rest/reference/git#create-a-reference
+                    // Updating the default branch of the fork to the sha results in:
+                    // `422 Object does not exist`
+                    // https://docs.github.com/enterprise/2.22/rest/reference/git#update-a-reference
+                    // Confirmed that this does not occur for another similarly old fork
+                    if (GitHub.isForkStale(parent, fork)) {
+                        log.info("Fork's commit graph is inconsistent," +
```
[the remark this answers is not in the document]
Comment (justinharringa) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java]: uh no, that's weird

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java [hunk 4]
```diff
@@ -48,7 +49,20 @@ public GHRepository getOrCreateFork(GHRepository parent) {
...
+                    // Check to see if latest sha from parent is in the tree of the fork. If not,
+                    // it is stale. If stale, an administrator will need to rebuild the commit
+                    // graph. If they don't do that, we get a 422 Reference update failed
+                    // https://docs.github.com/enterprise/2.22/rest/reference/git#create-a-reference
+                    // Updating the default branch of the fork to the sha results in:
+                    // `422 Object does not exist`
+                    // https://docs.github.com/enterprise/2.22/rest/reference/git#update-a-reference
+                    // Confirmed that this does not occur for another similarly old fork
+                    if (GitHub.isForkStale(parent, fork)) {
+                        log.warn("Fork's commit graph is inconsistent," +
```
[the remark this answers is not in the document]
Comment (justinharringa) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java:61]: that's weird... you're drunk codeclimate

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/repository/GitHub.java [hunk 5]
```diff
@@ -1,10 +1,13 @@
 package com.salesforce.dockerfileimageupdate.repository;
 
 import com.google.common.collect.Multimap;
+import org.kohsuke.github.GHFileNotFoundException;
 import org.kohsuke.github.GHRepository;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
```
[the remark this answers is not in the document]
Comment (justinharringa) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/repository/GitHub.java:9]: IntelliJ begs to differ. Not sure that this is important

[the remark this answers is not in the document]
Comment (justinharringa) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/repository/GitHub.java:9]: IntelliJ begs to differ. Not sure that this is `import`ant

File: dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (justinharringa) [reply dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/DockerfileGitHubUtil.java]: Done in f354e3c

**Review by afalko — APPROVED** [review]

**Comment by afalko** [comment]
@justinharringa Looks like code coverage dropped a teeny bit. Want to fix that, or I can merge despite that?

**Comment by justinharringa** [comment]
> @justinharringa Looks like code coverage dropped a teeny bit. Want to fix that, or I can merge despite that?

probably good to merge - static loggers like this are tricker to unit test - not sure if there are newer, nicer patterns for teeing up loggers like this while also making it simple to test them

**Review by afalko — APPROVED** [review]

**Pull request merged**